perftest/PerfTestSocketHandler.cpp

```diff
@@ -0,0 +1,188 @@
+    LOG_DBG("PerfTestSocketHandler done");
+}
+
+int PerfTestSocketHandler::getPollEvents(std::chrono::steady_clock::time_point now, int64_t &timeoutMaxMicroS)
+{
+    LOG_TRC("PerfTestSocketHandler getPollEvents");
+    if (_connecting) {
+        LOG_TRC("PerfTestSocketHandler getPollEvents (Waiting for outbound connection)");
+        return POLLOUT;
+    }
+
+    while (!_outgoingMessages.empty()) {
+        std::string message = _outgoingMessages.front();
+        _outgoingMessages.pop();
+        LOG_TRC("PerfTestSocketHandler Sending mesage: " << message);
+        WebSocketHandler::sendMessage(message);
+    }
+
+    if (_shutdown) {
+        LOG_DBG("PerfTestSocketHandler shutdown1");
+        WebSocketHandler::shutdown();
+    }
+
+    int events = WebSocketHandler::getPollEvents(now, timeoutMaxMicroS);
+    timeoutMaxMicroS = std::min(timeoutMaxMicroS, (int64_t)1000);
+    return events;
+}
+
+void PerfTestSocketHandler::handleMessage(const std::vector<char> &data)
+{
+    const std::string firstLine = COOLProtocol::getFirstLine(data.data(), data.size());
+    LOG_TRC("handleMessage: " + firstLine);
+
+    StringVector tokens = StringVector::tokenize(firstLine);
+    if (tokens.equals(0, "error:")) {
+        abort("Incoming error mesage: " + firstLine);
+    }
+
+    if (tokens.equals(0, "tile:")) {
+        // eg. tileprocessed tile=0:9216:0:3072:3072:0
+        TileDesc desc = TileDesc::parse(tokens);
+        WebSocketHandler::sendMessage("tileprocessed tile=" + desc.generateID());
+    }
+
+    _incomingMessages.emplace(data.begin(), data.end());
+}
+
+void PerfTestSocketHandler::abort(const std::string &message)
+{
+    std::cerr << "PerfTestSocketHandler abort: " << message << std::endl;
+    LOG_ERR("PerfTestSocketHandler abort: " << message);
+    WebSocketHandler::shutdown();
+    Util::forcedExit(EX_SOFTWARE);
+}
+
+MessagePerfTestSocketHandler::MessagePerfTestSocketHandler(
+        const std::string &name,
+        const std::string &server,
+        const std::string &messageFile,
+        std::atomic<bool>* measuring,
+        std::atomic<unsigned int>* messageCount,
+        std::atomic<unsigned int>* messageBytes,
+        std::atomic<unsigned int>* messageCountTile,
+        std::atomic<unsigned int>* messageBytesTile) :
+    PerfTestSocketHandler(name, server),
+    _messageFile(messageFile),
+    _measuring(measuring),
+    _messageCount(messageCount),
+    _messageBytes(messageBytes),
+    _messageCountTile(messageCountTile),
+    _messageBytesTile(messageBytesTile)
+{
+}
+
+void MessagePerfTestSocketHandler::handleMessage(const std::vector<char> &data)
+{
+    PerfTestSocketHandler::handleMessage(data);
+    if (*_measuring) {
+        const std::string firstLine = COOLProtocol::getFirstLine(data.data(), data.size());
+        StringVector tokens = StringVector::tokenize(firstLine);
+        if (tokens.equals(0, "tile:") || tokens.equals(0,"invalidatetiles:")) {
+            (*_messageCountTile)++;
```

Comment:
That is really odd C++ compared with how the Replay.hpp manages a Stats class; this is quite 'different'. A more normal C++ idiom would be to use a reference with '&' and get rid of the (* ) - around the member.